Initial Docs (#64)
janus-idp/operator: docs/admin.md

- By default each newly created Backstage CR uses Operator scope Default Configuration
- Which can be fully or partially overriden for particular CR instance using ConfigMap with witn name pointed in BackstageCR.spec.RawConfig 
- Which in turn can be customized by other BackstageCR.spec fields (see Backstage API doc)

Cluster Administrator may want to customize Default Configuration due to internal preferences/limitations, for example:
- Preferences/restrictions for Backstage and|or PostgreSQL images due to  Airgapped environment.
- Existed Taints and tolerations policy, so Backstage Pods have to be configured with certain tolerations restrictions.
- ...

Default Configuration is implemented as a ConfigMap called *backstage-default-config*, deployed on *backstage-system* namespace and mounted to Backstage controller container as a */default-config* directory.
This config map contains the set of keys/values which maps to file names/contents in the */default-config*.
These files contain yaml manifests of objects used by Backstage controller as an initial desired state of Backstage CR according to Backstage Operator data model.

> (TODO: link to the diagram here) 

Reviewer: part of this PR or a followup one?

Author: @nickboldt 
I do not know, depends on this PR dynamic

Mapping of configMap keys (yaml files) to runtime objects (NOTE: for the time (Dec 20'23) it is a subject of change):
| Key/File name         | k8s/OCP Kind        | Mandatory*    |Notes                                      |
| ----------------------|:-------------------:| --------------|------------------------------------------:|
| deployment.yaml       | appsv1.Deployment   | Yes           | Backstage deployment |
| service.yaml          | corev1.Service      | Yes           | Backstage Service |
| db-statefulset.yaml   | appsv1.Statefulset  | For DB enabled| PostgreSQL statefulSet    |    
| db-service.yaml       | corev1.Service      | For DB enabled| PostgreSQL Service   |
| db-secret.yaml        | corev1.Secret       | For DB enabled| Secret to connect Backstage to PSQL   |
| route.yaml            | openshift.Route     | No (for OCP)  | Route exposing Backstage service    |

> | app-config.yaml       | corev1.ConfigMap    | No            | Backstage app-config.yaml    |

Reviewer 2: But to me (even if it is subject to changes), this should strive to reflect the current state, no? And it would need to be updated accordingly in #70..
Because that's what the Docs team will leverage to begin with.

> | app-config.yaml       | corev1.ConfigMap    | No            | Backstage app-config.yaml    |

Author: Yes, the goal is to reflect #70

> | app-config.yaml       | corev1.ConfigMap    | No            | Backstage app-config.yaml    |

Reviewer 2: Can we make the two PRs (#64 and #70) independent?
The focus right now is to allow QE and docs teams to get started as soon as possible with what we have already, and a minimum of documentation is needed by the docs team (and soon other users).
So to me, this whole table could be removed from this PR and moved to #70 instead. WDYT?

This will also allow us to merge #64 as soon as possible and fix https://github.com/janus-idp/operator/issues/108

> | app-config.yaml       | corev1.ConfigMap    | No            | Backstage app-config.yaml    |

Author: Well, I would not like us support several versions of documents at this stage (before GA)
IMO it is quite labor intensive for near zero profit.

But, OK, I added one more "ver" field to give an idea (HTH temporarily)

> | app-config.yaml       | corev1.ConfigMap    | No            | Backstage app-config.yaml    |

Reviewer 2: Alright, sounds good then. Thanks.

| configmap-files.yaml  | corev1.ConfigMap    | No            | Backstage config file inclusions from configMap   |
| configmap-envs.yaml   | corev1.ConfigMap    | No            | Backstage env variables from configMap    |
| secret-files.yaml     | corev1.Secret       | No            | Backstage config file inclusions from Secret   |
| secret-envs.yaml      | corev1.Secret       | No            | Backstage env variables from Secret    |
| dynamic-plugins.yaml  | corev1.ConfigMap    | No            | dynamc-plugins config *    |

NOTES: 
 - Mandatory means it is needed to be present in either (or both) Default and CR Raw Configuration.
 - dynamic-plugins.yaml is a fragment of app-config.yaml provided with RHDH/Janus-IDP. The reason it mentioned separately is specific way it provided (via mounting to dedicated initContainer)  

### Operator Bundle configuration 

With Backstage Operator's Makefile you can generate bundle descriptor using *make bundle* command

Along with CSV manifest it generates default-config ConfigMap manifest, which can be modified and applied to Backstage Operator
TODO: is kubectl sufficient or there are some other tools?

> ### Kustomize deploy configuration

Reviewer: intentional spelling here? is there an openshift equivalent tool for downstreaming this text, like using `oc` instead of `kustomize` ?

Author: Nothing I am aware about. Please let me know if there is an equvalent.

Reviewer: so running `make deploy` will run `kustomize` ?

Author: yes

Make sure use the current context in your kubeconfig file is pointed to correct place, change necessary part of your config/manager/default-config or just replace some of the file(s) with yours and run
``
make deploy
``

### Direct ConfigMap configuration

You can change default configuration by directly changing the default-config ConfigMap with kubectl like:

>  - if you need to get existed default-config

Reviewer: ```suggestion
 - if you need to get the current default-config
```

``
kubectl get -n backstage-system configmap default-config > my-config.yaml
``

> - modify any part of it (yes, it could be fairly big file)

Reviewer: ```suggestion
- modify the file in your editor of choice
```

> - apply it back

Reviewer: ```suggestion
- apply the updated configuration to your cluster
```

``
  kubectl apply -n backstage-system -f my-config.yaml
``

> It has to be re-applied to the controller's container after some time.

Reviewer: "after some time" ... how long? why? do you mean it'll be overwritten when the operator subscription is updated from 1.1 -> 1.2?

Author: when k8s reconciles it.

### Use Cases

#### Airgapped environment

Creating Backtage CR, the Operator will try to create Backstage Pod, deploying:
- Backstage Container from the image, configured in *(deployment.yaml).spec.template.spec.Containers[].image*
- Init Container (applied for RHDH/Janus-IDP configuration, usually the same as Backstage Container)
- 
Also, if Backstage CR configured as EnabledLocalDb it will create Database (PGSQL) Container, configured in *(db-deployment.yaml).spec.template.spec.Containers[].image*

> By default Backstage Operator is configured to use publicly availabe images, which is not acceptable for Airgapped environment.

Reviewer: ```suggestion
By default Backstage Operator is configured to use publicly available images, which is not acceptable for Airgapped environment.
```

> By default Backstage Operator is configured to use publicly availabe images, which is not acceptable for Airgapped environment.

Reviewer: "not acceptable for Airgapped environment"

This is an unacceptable solution. We can't just say "if you want postgres, don't use airgap".

Instead we need a solution that will prefetch the relevant container inside the cluster.... https://access.redhat.com/documentation/en-us/red_hat_openshift_dev_spaces/3.10/html-single/administration_guide/index#installing-devspaces-in-a-restricted-environment ... and then use that same image and pinned digest when deploying the postgres pod.

Reviewer: see also https://github.com/janus-idp/operator/pull/86 which seeks to provide and OOTB airgap solution

Reviewer 2: I just noticed this unresolved comment. Can we rephrase this, e.g. to highlight the `RELATED_IMAGE_*` environment variables?

Author: I do not think I tried to say "if you want postgres, don't use airgap", it is just stated that default config uses publicly available images (which does not seem to be a problem?)

As discussed this topic is not finished and has to be carefully [tested and documented](https://github.com/janus-idp/operator/issues/80#issuecomment-1864733601) .

Reviewer: suggested some improved copy in https://github.com/janus-idp/operator/pull/64#discussion_r1439691147

# Administrator Guide

## Backstage Operator configuration

### Context

As it is described in Design doc (TODO), Backstage CR's desired state is defined using layered configuration approach, which means:
- By default each newly created Backstage CR uses Operator scope Default Configuration
- Which can be fully or partially overriden for particular CR instance using ConfigMap with the name pointed in BackstageCR.spec.RawConfig 
- Which in turn can be customized by other BackstageCR.spec fields (see Backstage API doc)

Cluster Administrator may want to customize Default Configuration due to internal preferences/limitations, for example:

> - Preferences/restrictions for Backstage and|or PostgreSQL images due to  Airgapped environment.

Reviewer: ```suggestion
- Preferences or restrictions for Backstage;
- A PostgreSQL image that is available from within a restricted (air-gapped) environment;
```

Author: What I am trying to say is the  Backstage and|or PostgreSQL IMAGES can be a subject of customization of Backstage config.

Reviewer: ok, revised my suggestion. is that closer to what you were trying to convey?

> - Existed Taints and tolerations policy, so Backstage Pods have to be configured with certain tolerations restrictions.

Reviewer: you have capitalized Taints but not Tolerations. Intentional?

Author: no, it's typo

| db-secret.yaml        | corev1.Secret       | For DB enabled|all |Secret to connect Backstage to PSQL   |
| route.yaml            | openshift.Route     | No (for OCP)  |all |Route exposing Backstage service    |
| app-config.yaml       | corev1.ConfigMap    | No            |0.02|Backstage app-config.yaml    |
| configmap-files.yaml  | corev1.ConfigMap    | No            |0.02|Backstage config file inclusions from configMap   |
| configmap-envs.yaml   | corev1.ConfigMap    | No            |0.02|Backstage env variables from configMap    |
| secret-files.yaml     | corev1.Secret       | No            |0.02|Backstage config file inclusions from Secret   |
| secret-envs.yaml      | corev1.Secret       | No            |0.02|Backstage env variables from Secret    |
| dynamic-plugins.yaml  | corev1.ConfigMap    | No            |0.02|dynamc-plugins config *    |
| dynamic-plugins-      | corev1.ConfigMap    | No            |0.01|dynamc-plugins config *    |
|configmap.yaml         |                     |               |    | |
| backend-auth-         | corev1.ConfigMap    | No            |0.01|backend auth config    |

> |configmap.yaml         |                     |               |    | |

Reviewer 2: The rendering of this table seems to have been lost with the latest changes:

![image](https://github.com/janus-idp/operator/assets/593208/f642d86a-10ce-47f6-aad3-cb9269f11310)

Can you fix it before t can be approved?

Reviewer 2: Formatting fixed in [`ebb6c8f` (#64)](https://github.com/janus-idp/operator/pull/64/commits/ebb6c8f9045c9306729e2d912abee326b5c804bb)

NOTES: 
 - Mandatory means it is needed to be present in either (or both) Default and CR Raw Configuration.

>  - dynamic-plugins.yaml is a fragment of app-config.yaml provided with RHDH/Janus-IDP. The reason it mentioned separately is specific way it provided (via mounting to dedicated initContainer)  

Reviewer: ```suggestion
 - dynamic-plugins.yaml is a fragment of app-config.yaml provided with RHDH/Janus-IDP, which is mounted into a dedicated initContainer. 
```

>  - items marked as version 0.01 are not supported in version 0.02 

Reviewer: This looks like an invalid semver version. Do you mean 0.0.1 -> 0.0.2?

### Operator Bundle configuration 

With Backstage Operator's Makefile you can generate bundle descriptor using *make bundle* command

Along with CSV manifest it generates default-config ConfigMap manifest, which can be modified and applied to Backstage Operator

> TODO: is kubectl sufficient or there are some other tools?

Reviewer: you mean like `oc` ?

Author: I mean we need document how to Admin is able to make change in default operator configuration.
He/she may have own configuration file (based on generated one) and apply it using kubectl or oc when needed.

### Kustomize deploy configuration

Make sure use the current context in your kubeconfig file is pointed to correct place, change necessary part of your config/manager/default-config or just replace some of the file(s) with yours and run
``
make deploy
``

### Direct ConfigMap configuration

You can change default configuration by directly changing the default-config ConfigMap with kubectl like:

 - retrieve the current `default-config` from the cluster

``
kubectl get -n backstage-system configmap default-config > my-config.yaml
``

- modify the file in your editor of choice

- apply the updated configuration to your cluster

``
  kubectl apply -n backstage-system -f my-config.yaml
``

It has to be re-applied to the controller's container after being reconciled by kubernetes processes.

### Use Cases

#### Airgapped environment

> Creating Backtage CR, the Operator will try to create Backstage Pod, deploying:

Reviewer: ```suggestion
When creating the Backstage CR, the Operator will try to create a Backstage Pod, deploying:
```

- Backstage Container from the image, configured in *(deployment.yaml).spec.template.spec.Containers[].image*
- Init Container (applied for RHDH/Janus-IDP configuration, usually the same as Backstage Container)
- 

> Also, if Backstage CR configured as EnabledLocalDb it will create Database (PGSQL) Container, configured in *(db-deployment.yaml).spec.template.spec.Containers[].image*

Reviewer: ```suggestion
Also, if Backstage CR configured with *EnabledLocalDb*,  it will create a PostgreSQL container pod, configured in *(db-deployment.yaml).spec.template.spec.Containers[].image*
```

> By default Backstage Operator is configured to use publicly available images, which is not acceptable for Airgapped environment.

Reviewer: ```suggestion
By default the Backstage Operator is configured to use publicly available images from registry.redhat.io or quay.io. If you plan to deploy to a restricted environment, you will need to configure your cluster or network to allow these images to be pulled. For the list of related images deployed by the Operator, see the `RELATED_IMAGE` env vars or `relatedImages` section of the [CSV](https://github.com/janus-idp/operator/blob/main/bundle/manifests/backstage-operator.clusterserviceversion.yaml). 
```

Author: I do not mind of removing "which is not acceptable for Airgapped" and your additions are good but I'd prefer to structure the document like:
- generic part (no dependency on registry.redhat.io or quay.io)
- OLM specific
- Openshift specific
- RedHat infrastructure specific

Reviewer: * How can you have zero dependency on quay.io, given both the helmchart and operator deploy images from quay.io?  Are you talking about the process for mirroring into your cluster and then using those mirrored images within an airgapped / restricted environment?

* I don't understand the distinction between OLM and Openshift, since OLM is only supported on OpenShift. Or can it be used on vanilla K8S too?

* I don't see a mention of RH infrastructure. Are you talking about Openshift's implementation of restricted envinronment (air gap) support ?

| secret-envs.yaml               | corev1.Secret      | No             | 0.0.2   | Backstage env variables from Secret             |
| dynamic-plugins.yaml           | corev1.ConfigMap   | No             | 0.0.2   | dynamic-plugins config *                        |
| dynamic-plugins-configmap.yaml | corev1.ConfigMap   | No             | 0.0.1   | dynamic-plugins config *                        |
| backend-auth-configmap.yaml    | corev1.ConfigMap   | No             | 0.0.1   | backend auth config                             |

NOTES: 
 - Mandatory means it is needed to be present in either (or both) Default and CR Raw Configuration.
 - dynamic-plugins.yaml is a fragment of app-config.yaml provided with RHDH/Janus-IDP, which is mounted into a dedicated initContainer. 
 - items marked as version 0.0.1 are not supported in version 0.0.2 
### Operator Bundle configuration 

With Backstage Operator's Makefile you can generate bundle descriptor using *make bundle* command

Along with CSV manifest it generates default-config ConfigMap manifest, which can be modified and applied to Backstage Operator.

> [//]: # (TODO: is kubectl sufficient or there are some other tools?)

Reviewer: ```suggestion
[//]: # (TODO: document how an administrator can make changes to the default operator configuration, using their own configuration file (perhaps based on the generated one), and apply it using `kubectl` or `oc`.
```
